### ISSUE TITLE
provide a method to shrink the storage of an array

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#2879]
After growing an array, it may have up to twice the necessary storage. It would be nice, in addition to `sizehint` to have some function to "shrinkwrap" a collection. One possible API would be to just call `sizehint(v,length(v))`.

Original proposal of this idea on julia-dev: https://groups.google.com/forum/?fromgroups=#!search/sizehint/julia-dev/ryZLQTtjJKc/bRbAyh7T6dMJ.

**Comment by holocronweaver** [comment]
Has this been implemented yet, or is it still an open feature request? Anyone know the answers to questions by @quinnj?

**Comment by StefanKarpinski** [comment]
A decent API for this would be to call `sizehint!(v, sz)` with a value of `sz` that ≤ the actual size of `v`.

**Comment by sbromberger** [comment]
I could come up with a PR if people are interested...

**Comment by tkelman** [comment]
Try it out locally, see if it looks like it makes things behave as you'd like? Sounds worth a PR to me.

**Comment by sbromberger** [comment]
Is there a way within Julia to see how much memory has been allocated for, say, a Vector{Int}() that has been `sizehint!()`ed?

**Comment by sbromberger** [comment]
Actually, on second thought, sizehint!() doesn't do anything to memory allocation: with some debugging info printed,
```
julia> a = Vector{Int}()
0-element Array{Int64,1}

julia> sizehint!(a,9_000_000_000_000)
array has grown by 9000000000000 bytes; now 9000000000000
array maxsize = 9000000000000, nrows = 9000000000000
0-element Array{Int64,1}
```
... and Julia's memory utilization (as seen in activity monitor) remains unchanged.

**Comment by yuyichao** [comment]
Malloc/mmap is lazy and won't fault the page unless you access it

**Comment by StefanKarpinski** [comment]
For large uninitialized arrays, the kernel gives you pages lazily – i.e. only when you write to them.

**Comment by sbromberger** [comment]
Thanks. So this means, then, that overestimating `sizehint!()` by a huge factor won't actually cause any significant overallocation.

**Comment by JeffBezanson** [comment]
Closed by #26201.
